README
CPAN::Distribution::ReleaseHistory



INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for CPAN-Distribution-ReleaseHistory

0.002006 2017-03-06T13:51:33Z c0cfab8

0.002005 2017-03-06T13:12:15Z e869b9f
 [Bugfix]
 - Fix tests failing with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 0.002004, see misc/*.deps* for details
 - develop: ↑1 (suggests: ↑2)
 - test: ↓1

0.002004 2016-11-08T09:19:00Z 4349a15
 [Bugfix / #1]
 - Fix interoperability with new ES 5.0 Client libraries.
 - Closes https://github.com/kentnl/CPAN-Distribution-ReleaseHistory/issues/1
 - Thanks to Slaven Rezić for report

 [Dependencies::Stats]
 - Dependencies changed since 0.002003, see misc/*.deps* for details
 - configure: (recommends: ↓1)
 - develop: +10 ↑3 -2 (recommends: ↓1, suggests: ↑2)
 - runtime: -1
 - test: (recommends: ↓2)

0.002003 2014-11-12T16:48:51Z 1c14e8a
 [Dependencies::Stats]
 - Dependencies changed since 0.002002, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: (suggests: ↑1)
 - test: +1 (recommends: ↑1)

 [Tests]
 - Further improve tests to spew things when they fail, esp: Avoid crappy subtests that give no context.

0.002002 2014-11-08T02:54:43Z 0c2c049
 [Dependencies::Stats]
 - Dependencies changed since 0.002001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +6 ↑1 -1 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1
 - test: +2 (recommends: ↑2)

 [Metadata]
 - author = kentnl

 [Tests]
 - Use Test::RequiresInternet to shore up test failures.

0.002001 2014-08-16T07:19:26Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests updated
 - metadata updated.
 - whitespace reduced.

 [Dependencies::Stats]
 - Dependencies changed since 0.002000, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 ↑1 (suggests: ↑2)
 - test: +1 -3 (recommends: ↑1 -1)

 [Metadata]
 - migrated to kentnl org.

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - Update Test::ReportPrereqs

0.002000 2014-06-12T20:47:43Z
 [00 API Impacting changes]
 - Code refactored internally to allow better overriding in subclasses
 - ::ReleaseIterator now takes 'scroller' attribute instead.

 [Documentation]
 - misc spurious references to CPAN::Latest and MetaCPAN::Client removed.

 [Examples]
 - new examples/dist-days.pl
 - new examples/dist-hours.pl
 - new examples/dist-months.pl
 - new examples/dist-releases.pl
 - new examples/latest-dist-release.pl

0.001000 2014-06-12T16:01:13Z
 - First version.


examples/dist-hours.pl
#!/usr/bin/env perl
# ABSTRACT: Show day-of-week distribution for a given dist

use strict;
use warnings;
use utf8;

use lib 'lib';

use CPAN::Distribution::ReleaseHistory;

die "Specify a distribution name" unless my $distname = shift @ARGV;

my $rh = CPAN::Distribution::ReleaseHistory->new( distribution => $distname );
my $iterator = $rh->release_iterator;

my %buckets = ( map { $_ => 0 } 0 .. 23 );

sub record_timestamp {
  my ($ts)     = @_;
  my (@fields) = gmtime $ts;
  $buckets{ $fields[2] }++;
}
while ( my $release = $iterator->next_release ) {
  record_timestamp( $release->timestamp );
  printf "%s\t%s\n", $release->timestamp, $release->path;
}
my $max = [ sort { $a <=> $b } values %buckets ]->[-1];
my $scale = 1;
$scale = 40 / $max if $max > 40;
for my $key ( sort { $a <=> $b } keys %buckets ) {
  printf "%s : %s\n", $key, "#" x ( int( $buckets{$key} * $scale ) );
}



examples/dist-releases.pl
#!/usr/bin/env perl
# FILENAME: dist-releases.pl
# CREATED: 06/13/14 06:46:31 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Show all the releases of a given dist

use strict;
use warnings;
use utf8;

use lib 'lib';

use CPAN::Distribution::ReleaseHistory;

die "Specify a distribution name" unless my $distname = shift @ARGV;

my $rh = CPAN::Distribution::ReleaseHistory->new( distribution => $distname );
my $iterator = $rh->release_iterator;
my $last;
my @intervals;
while ( my $release = $iterator->next_release ) {
  my $delta = '';
  if ($last) {
    $delta = $last - $release->timestamp;
    push @intervals, $delta;
  }
  $last = $release->timestamp;
  printf "%s\t%s\n", $release->timestamp, $release->path;
}

my @sorted = sort { $a <=> $b } @intervals;

sub format_delta {
  my ($delta) = @_;
  my $string = '';

  if ( $delta < 1 ) {
    return '0s';
  }
  $string = ( $delta % 60 ) . 's';
  $delta  = int( $delta / 60 );
  return $string if $delta < 1;
  $string = ( $delta % 60 ) . 'm ' . $string;
  $delta  = int( $delta / 60 );
  return $string if $delta < 1;
  $string = ( $delta % 24 ) . 'h ' . $string;
  $delta  = int( $delta / 24 );
  return $string if $delta < 1;
  return $delta . 'd ' . $string;
}

print "# -- Interval Stats\n";
printf "#min           : %s\n", format_delta( $sorted[0] );
printf "#lower quartile: %s\n", format_delta( $sorted[ int( $#sorted / 4 ) ] );
printf "#median        : %s\n", format_delta( $sorted[ int( $#sorted / 2 ) ] );
printf "#upper quartile: %s\n", format_delta( $sorted[ int( $#sorted / 4 * 3 ) ] );
printf "#max           : %s\n", format_delta( $sorted[-1] );



lib/CPAN/Distribution/ReleaseHistory.pm
use 5.006;
use strict;
use warnings;

package CPAN::Distribution::ReleaseHistory;

our $VERSION = '0.002006';

# ABSTRACT: Show the release history of a single distribution

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo 1.000008 qw( has );














has 'distribution' => (
  is       => 'ro',
  required => 1,
);



































has 'sort' => (
  is      => 'ro',
  lazy    => 1,
  builder => sub { 'desc' },
);














has 'scroll_size' => (
  is      => 'ro',
  lazy    => 1,
  builder => sub { 1000 },
);












sub _iterator_from_scroll {
  my ( undef, $scroll ) = @_;
  require CPAN::Distribution::ReleaseHistory::ReleaseIterator;
  return CPAN::Distribution::ReleaseHistory::ReleaseIterator->new( scroller => $scroll );
}

sub release_iterator {
  my ($self) = @_;
  return $self->_iterator_from_scroll( $self->_mk_query_distribution );
}













has 'ua' => (
  is        => 'ro',
  predicate => 'has_ua',
);







has 'es' => (
  is      => 'ro',
  lazy    => 1,
  builder => sub {
    my ($self) = @_;
    my %args = (
      nodes            => 'api.metacpan.org',
      cxn_pool         => 'Static::NoPing',
      send_get_body_as => 'POST',
    );
    if ( $self->has_ua ) {
      $args{handle} = $self->ua;
    }
    require Search::Elasticsearch;
    return Search::Elasticsearch->new(%args);
  },
);

sub _mk_query {
  my ($self) = @_;
  return { term => { distribution => $self->distribution } };
}

sub _mk_body {
  my ($self) = @_;
  my $body = { query => $self->_mk_query };
  if ( $self->sort ) {
    $body->{sort} = { 'stat.mtime' => $self->sort };
  }
  return $body;
}

sub _mk_fields {
  return [qw(name version date status maturity stat download_url )];
}

sub _es_version {
  my ( $self, $wanted_version ) = @_;
  local $@ = undef;
  return eval { $self->es->VERSION($wanted_version); 1 };    ## no critic (RequireCheckingReturnValueOfEval)
}

sub _mk_scroll_args {
  my ($self) = @_;

  my %scrollargs = (
    scroll => '5m',
    index  => 'v0',
    type   => 'release',
    size   => $self->scroll_size,
    body   => $self->_mk_body,
    ( $self->_es_version(5) ? 'stored_fields' : 'fields' ) => $self->_mk_fields,
  );

  if ( not $self->sort ) {
    $scrollargs{'search_type'} = 'scan';
  }
  return \%scrollargs;
}

sub _mk_query_distribution {
  my ($self) = @_;

  my %scrollargs = %{ $self->_mk_scroll_args };

  ( $self->_es_version(5) )
    ? ( require Search::Elasticsearch::Client::5_0::Scroll )
    : ( require Search::Elasticsearch::Scroll );

  return $self->es->scroll_helper(%scrollargs);
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

CPAN::Distribution::ReleaseHistory - Show the release history of a single distribution

=head1 VERSION

version 0.002006

=head1 SYNOPSIS

This is similar in concept to C<CPAN::ReleaseHistory>, except its tailored to use a single distribution name, and uses
C<MetaCPAN> to resolve its information.

  use CPAN::Distribution::ReleaseHistory;

  my $release_history = CPAN::Distribution::ReleaseHistory->new(
    distribution => 'Dist-Zilla',
    # ua          => a HTTP::Tiny instance to use for requests
    # es          => a Search::Elasticsearch instance
    # scroll_size => 1000  : How many results to fetch per HTTP request
    # sort        => 'desc': Direction of sort ( vs 'asc' and undef )
  );

  # Returns a CPAN::Distribution::ReleaseHistory::ReleaseIterator
  my $iterator = $release_history->release_iterator();

  # $release is an instance of CPAN::Distribution::ReleaseHistory::Release
  while ( my $release = $iterator->next_release() ) {
    print $release->distname();                   # Dist-Zilla
    print $release->path();                       # R/RJ/RJBS/Dist-Zilla-1.000.tar.gz
    print scalar gmtime $release->timestamp();    # Timestamp is Unixtime.
    print $release->size();                       # 30470 ( bytes )
    my $distinfo = $release->distinfo();          # CPAN::DistInfo object
  }

=head1 METHODS

=head2 C<release_iterator>

Perform the query and return a new
L<< C<CPAN::Distribution::ReleaseHistory::ReleaseIterator>|CPAN::Distribution::ReleaseHistory::ReleaseIterator >> to walk over
the results.

  my $iterator = $object->release_iterator

=head2 C<has_ua>

Determine if user specified a custom C<UserAgent>

=head1 ATTRIBUTES

=head2 C<distribution>

A string exactly matching a name of a C<CPAN> distribution.

example:

  Dist-Zilla
  MetaCPAN-Client
  Search-Elasticsearch
  WWW-Mechanize-Cached

=head2 C<sort>

The implicit sort direction of the output.

  default: 'desc' # The most recent release is returned first.

Alternative options:

  'asc' # The oldest release is returned first
  undef # Results are unsorted

=head4 C<undef>

Opting for C<undef> for this value will give a slight speed up to the responsiveness of queries.

Though this benefit will only be observed in conjunction with low values of C<scroll_size>

    5 desc average 0.08625 /each   11.594 items/sec
   5 undef average 0.03856 /each   25.937 items/sec

   10 desc average 0.05384 /each   18.573 items/sec
  10 undef average 0.03773 /each   26.507 items/sec

   20 desc average 0.03856 /each   25.934 items/sec
  20 undef average 0.02758 /each   36.252 items/sec

   50 desc average 0.02579 /each   38.777 items/sec
  50 undef average 0.02547 /each   39.267 items/sec

  100 desc average 0.02279 /each   43.873 items/sec
 100 undef average 0.02510 /each   39.846 items/sec

=head2 C<scroll_size>

Volume of results to fetch per request.

  default: 1000

Larger values give slower responses but faster total execution time.

Smaller values give faster responses but slower total execution time. ( Due to paying ping time both ways per request in
addition to other per-request overheads that are constant sized )

=head2 C<ua>

A C<HTTP::Tiny> compatible user agent.

=head2 C<es>

A Search::Elasticsearch instance.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/CPAN/Distribution/ReleaseHistory/Release.pm
use 5.006;
use strict;
use warnings;

package CPAN::Distribution::ReleaseHistory::Release;

our $VERSION = '0.002006';

# ABSTRACT: A container for release data

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( has );

use CPAN::DistnameInfo;









has 'distname' => ( is => 'ro' );









has 'path' => ( is => 'ro' );







has 'timestamp' => ( is => 'ro' );







has 'size' => ( is => 'ro' );







has 'distinfo' => ( is => 'lazy' );

sub _build_distinfo {
  my $self = shift;

  return CPAN::DistnameInfo->new( $self->path );
}
no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

CPAN::Distribution::ReleaseHistory::Release - A container for release data

=head1 VERSION

version 0.002006

=head1 SYNOPSIS

This is mostly a work-a-like for
L<< C<CPAN::Latest::ReleaseHistory::Release>|CPAN::Latest::ReleaseHistory::Release >>, except without the dependency on
L<< C<MetaCPAN::Client>|MetaCPAN::Client >>

  my $release = $releaseiterator->next_release;

  print $release->distname();                   # Dist-Zilla
  print $release->path();                       # R/RJ/RJBS/Dist-Zilla-1.000.tar.gz
  print scalar gmtime $release->timestamp();    # Timestamp is Unixtime.
  print $release->size();                       # 30470 ( bytes )
  my $distinfo = $release->distinfo();          # CPAN::DistInfo object

=head1 ATTRIBUTES

=head2 C<distname>

The name of the distribution.

  e.g: Dist-Zilla

=head2 C<path>

The path to the distribution relative to a C<CPAN> mirror.

  e.g: R/RJ/RJBS/Dist-Zilla-1.000.tar.gz

=head2 C<timestamp>

The time of the release in C<unixtime>

=head2 C<size>

The size of the release in C<bytes>

=head2 C<distinfo>

A L<< C<CPAN::DistnameInfo>|CPAN::DistnameInfo >> object for this release.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/CPAN/Distribution/ReleaseHistory/ReleaseIterator.pm
use 5.006;
use strict;
use warnings;

package CPAN::Distribution::ReleaseHistory::ReleaseIterator;

our $VERSION = '0.002006';

# ABSTRACT: A container to iterate a collection of releases for a single distribution

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( has );
use CPAN::DistnameInfo;
use CPAN::Distribution::ReleaseHistory::Release;







has 'scroller' => ( is => 'ro', required => 1 );









sub next_release {
  my ($self) = @_;
  my $scroll_result = $self->scroller->next;
  return if not $scroll_result;

  my $data_hash = $scroll_result->{'_source'} || $scroll_result->{'fields'};
  return if not $data_hash;

  my $path = $data_hash->{download_url};
  $path =~ s{\A.*/authors/id/}{}msx;
  my $distinfo = CPAN::DistnameInfo->new($path);
  my $distname =
    defined($distinfo) && defined( $distinfo->dist )
    ? $distinfo->dist
    : $data_hash->{name};
  return CPAN::Distribution::ReleaseHistory::Release->new(
    distname  => $distname,
    path      => $path,
    timestamp => $data_hash->{stat}->{mtime},
    size      => $data_hash->{stat}->{size},
  );
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

CPAN::Distribution::ReleaseHistory::ReleaseIterator - A container to iterate a collection of releases for a single distribution

=head1 VERSION

version 0.002006

=head1 METHODS

=head2 C<next_release>

Returns a L<< C<CPAN::Distribution::ReleaseHistory::Release>|CPAN::Distribution::ReleaseHistory::Release >>

  my $item = $release_iterator->next_release();

=head1 ATTRIBUTES

=head2 C<scroller>

A C<Search::Elasticsearch::Scroll>  instance that dispatches results.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Show the release history of a single distribution",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "CPAN-Distribution-ReleaseHistory",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "CPAN::Distribution::ReleaseHistory",
  "PREREQ_PM" => {
    "CPAN::DistnameInfo" => 0,
    "Moo" => "1.000008",
    "Search::Elasticsearch" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0,
    "Test::RequiresInternet" => 0
  },
  "VERSION" => "0.002006",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::DistnameInfo" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moo" => "1.000008",
  "Search::Elasticsearch" => 0,
  "Test::Fatal" => 0,
  "Test::More" => 0,
  "Test::RequiresInternet" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
